fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/InliningStorageAdapter.java

```diff
@@ -0,0 +1,372 @@
+        onReadListener.onNodeRead(layer, node);
+        return node;
+    }
+
+    /**
+     * Constructs a {@code NodeReferenceWithVector} from raw key and value byte arrays retrieved from storage.
+     * <p>
+     * This helper method deserializes a neighbor's data. It unpacks the provided {@code key} to extract the neighbor's
+     * primary key and unpacks the {@code value} to extract the neighbor's vector. It also notifies the configured
+     * {@link OnReadListener} of the read operation.
+     *
+     * @param storageTransform an affine vector transformation operator that is used to transform the fetched vector
+     *        into the storage space that is currently being used
+     * @param layer the layer of the graph where the neighbor node is located.
+     * @param key the raw byte array key from the database, which contains the neighbor's primary key.
+     * @param value the raw byte array value from the database, which represents the neighbor's vector.
+     * @return a new {@link NodeReferenceWithVector} instance representing the deserialized neighbor.
+     * @throws IllegalArgumentException if the key or value byte arrays are malformed and cannot be unpacked.
+     */
+    @Nonnull
+    private NodeReferenceWithVector neighborFromRaw(@Nonnull final AffineOperator storageTransform, final int layer,
+                                                    @Nonnull final byte[] key, @Nonnull final byte[] value) {
+        final OnReadListener onReadListener = getOnReadListener();
+        onReadListener.onKeyValueRead(layer, key, value);
+
+        final Tuple neighborKeyTuple = getDataSubspace().unpack(key);
+        final Tuple neighborValueTuple = Tuple.fromBytes(value);
+
+        return neighborFromTuples(storageTransform, neighborKeyTuple, neighborValueTuple);
+    }
+
+    /**
+     * Constructs a {@code NodeReferenceWithVector} from tuples retrieved from storage.
+     * <p>
+     * @param storageTransform an affine vector transformation operator that is used to transform the fetched vector
+     *        into the storage space that is currently being used
+     * @param keyTuple the key tuple from the database, which contains the neighbor's primary key.
+     * @param valueTuple the value tuple from the database, which represents the neighbor's vector.
+     * @return a new {@link NodeReferenceWithVector} instance representing the deserialized neighbor.
+     * @throws IllegalArgumentException if the key or value byte arrays are malformed and cannot be unpacked.
+     */
+    @Nonnull
+    private NodeReferenceWithVector neighborFromTuples(@Nonnull final AffineOperator storageTransform,
+                                                       @Nonnull final Tuple keyTuple, @Nonnull final Tuple valueTuple) {
+        final Tuple neighborPrimaryKey = keyTuple.getNestedTuple(2); // neighbor primary key
+        final Transformed<RealVector> neighborVector =
+                storageTransform.transform(
```

Comment:
It actually confused me a bit why you were calling `transform` here. I believe it's correct, but it hinges on knowing that the underlying vector is only semantically modified if it is not an `EncodedVector`. (And also, the `storageTransform` is always the identity transformation if we are not using RaBitQ.) So it means that if we stored the non-transformed data (for example, we hadn't enabled RaBitQ on the HNSW yet), then this will transform the underlying data at read time. But if we had already done that, then we just keep what we read from storage.

It makes it seem more like `ensureTransformed` or something. Maybe that's a bit more obvious if this is pushed into the `StorageAdapter`. It's also possible that this is just tricky and there's not much we can do

Reply:
Your assessment is exactly right. I don't think I can make this more concise but I will leave a comment in the code explaining this better.